sim/encounters/dynamic_adds_ai.go

```diff
@@ -0,0 +1,235 @@
+
+			Stats: stats.Stats{
+				stats.Armor:       24835,
+				stats.AttackPower: 0,
+			}.ToProtoArray(),
+
+			SpellSchool:   proto.SpellSchool_SpellSchoolPhysical,
+			SwingSpeed:    2.0,
+			MinBaseDamage: 900000,
+			DamageSpread:  0.5,
+			TargetInputs:  dynamicAddsTargetInputs(),
+		},
+
+		AI: makeDynamicAddsAI(true),
+	})
+
+	core.AddPresetTarget(&core.PresetTarget{
+		PathPrefix: "Default",
+
+		Config: &proto.Target{
+			Id:        dynamicAddID,
+			Name:      addName,
+			Level:     93,
+			MobType:   proto.MobType_MobTypeMechanical,
+			TankIndex: 1,
+
+			Stats: stats.Stats{
+				stats.Armor:       24835,
+				stats.AttackPower: 0,
+			}.ToProtoArray(),
+
+			SpellSchool:   proto.SpellSchool_SpellSchoolPhysical,
+			SwingSpeed:    1.5,
+			MinBaseDamage: 300000,
+			DamageSpread:  0.4,
+			TargetInputs:  []*proto.TargetInput{},
+		},
+
+		AI: makeDynamicAddsAI(false),
+	})
+
+	core.AddPresetEncounter("Dynamic", []string{
+		"Default/" + bossName,
+		"Default/" + addName,
+	})
+}
+
+func dynamicAddsTargetInputs() []*proto.TargetInput {
+	return []*proto.TargetInput{
+		{
+			Label:       "Add(s) respawn Time",
+			Tooltip:     "Time for add(s) to respawn after previous died (in seconds)",
+			InputType:   proto.InputType_Number,
+			NumberValue: 30,
+		},
+		{
+			Label:       "Add(s) lifetime",
+			Tooltip:     "How long the add(s) stay alive (in seconds)",
+			InputType:   proto.InputType_Number,
+			NumberValue: 20,
+		},
+		{
+			Label:       "Add(s) spawn delay",
+			Tooltip:     "Initial delay before the add(s) spawn (in seconds)",
+			InputType:   proto.InputType_Number,
+			NumberValue: 10,
+		},
+	}
+}
+
+func makeDynamicAddsAI(isBoss bool) core.AIFactory {
+	return func() core.TargetAI {
+		return &DynamicAddsAI{
+			isBoss: isBoss,
+		}
+	}
+}
+
+type DynamicAddsAI struct {
+	Target   *core.Target
+	BossUnit *core.Unit
+	AddUnits []*core.Unit
+	MainTank *core.Unit
+	OffTank  *core.Unit
+
+	// Static parameters associated with a given preset
+	isBoss bool
+
+	respawnTime time.Duration
+	addLifetime time.Duration
+	spawnDelay  time.Duration
+
+	nextSpawnTime time.Duration
+}
+
+func (ai *DynamicAddsAI) Initialize(target *core.Target, config *proto.Target) {
+	ai.Target = target
+	ai.Target.AutoAttacks.MHConfig().ActionID.Tag = core.TernaryInt32(ai.isBoss, dynamicBossID, dynamicAddID)
+
+	if ai.isBoss {
+		ai.BossUnit = &target.Unit
+		for _, encounterTarget := range target.Env.Encounter.AllTargetUnits {
+			if encounterTarget != &target.Unit {
+				ai.AddUnits = append(ai.AddUnits, encounterTarget)
+			}
+		}
+	} else {
+		for _, encounterTarget := range target.Env.Encounter.AllTargets {
+			if encounterTarget.AI != nil {
+				if bossAI, ok := encounterTarget.AI.(*DynamicAddsAI); ok && bossAI.isBoss {
+					ai.BossUnit = &encounterTarget.Unit
+					break
+				}
+			}
+		}
+	}
+
+	if ai.BossUnit != nil {
+		ai.MainTank = ai.BossUnit.CurrentTarget
+	}
+	if len(ai.AddUnits) > 0 {
+		ai.OffTank = ai.AddUnits[0].CurrentTarget
+	}
```

Comment:
Likewise, you can just assume that there is at least 1 add unit always configured, or else the AI doesn't make any sense.